sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuGenerateExec.scala

```diff
@@ -36,75 +38,345 @@ class GpuGenerateExecSparkPlanMeta(
+      vectors(generatorOffset).getChildColumnView(0)) { listValues =>
+      val totalSize = listValues.getDeviceMemorySize
+      val totalCount = listValues.getRowCount
+      (totalSize.toDouble, totalCount.toDouble)
+    }
+    // input size of columns to be repeated during exploding
+    val repeatColsInputSize = vectors.slice(0, generatorOffset).map(_.getDeviceMemorySize).sum
+    // estimated output size of repeated columns
+    val repeatColsOutputSize = repeatColsInputSize * estimatedOutputRows / inputRows
+    // estimated total output size
+    val estimatedOutputSizeBytes = explodeColOutputSize + repeatColsOutputSize
+    // how may splits will we need to keep the output size under the target size
+    val numSplitsForTargetSize = math.ceil(estimatedOutputSizeBytes / targetSizeBytes).toInt
+    // how may splits will we need to keep the output rows under max value
+    val numSplitsForTargetRow = math.ceil(estimatedOutputRows / Int.MaxValue).toInt
+    // how may splits will we need to keep exploding working safely
+    val numSplits = numSplitsForTargetSize max numSplitsForTargetRow
+
+    if (numSplits == 0) Array()
+    else GpuBatchUtils.generateSplitIndices(inputRows, numSplits)
+  }
+
+  // Infer result schema of GenerateExec from input schema
+  protected def resultSchema(inputSchema: Array[DataType],
+    genOffset: Int,
+    includePos: Boolean = false): Array[DataType] = {
+    val outputSchema = ArrayBuffer[DataType]()
+    inputSchema.zipWithIndex.foreach {
+      // extract output type of explode from input ArrayData
+      case (dataType, index) if index == genOffset =>
+        require(dataType.isInstanceOf[ArrayType], "GpuExplode only supports ArrayData now")
+        if (includePos) {
+          outputSchema += IntegerType
+        }
+        outputSchema += dataType.asInstanceOf[ArrayType].elementType
+      // map types of other required columns
+      case (dataType, _) =>
+        outputSchema += dataType
+    }
+    outputSchema.toArray
+  }
+
+  override def fixedLenLazyExpressions: Seq[Expression] = child match {
+    // GpuLiteral of ArrayData will be converted to GpuCreateArray with GpuLiterals
+    case GpuCreateArray(expressions, _) => expressions
+    case GpuAlias(GpuCreateArray(expressions, _), _) => expressions
+    case _ => Nil
+  }
+
+  override def fixedLenLazyArrayGenerate(inputIterator: Iterator[ColumnarBatch],
+      boundLazyProjectList: Seq[Expression],
+      boundOthersProjectList: Seq[Expression],
+      outputSchema: Array[DataType],
+      outer: Boolean,
+      numOutputRows: GpuMetric,
+      numOutputBatches: GpuMetric,
+      totalTime: GpuMetric): Iterator[ColumnarBatch] = {
+
+    val numArrayColumns = boundLazyProjectList.length
+    val numOtherColumns = boundOthersProjectList.length
+    val numExplodeColumns = if (position) 2 else 1
+
+    new Iterator[ColumnarBatch] {
+
+      var currentBatch: ColumnarBatch = _
```

Comment:
nit: can we have a follow on issue filed to make this spillable? I don't want it to happen here, but we should at least be tracking it.

Reply:
Done.